src/Microsoft.DotNet.Interactive.Formatting.Tests/FormatterTests.cs

```diff
@@ -211,468 +197,538 @@ public void Check_default_HTML_formatting_of_unicode()
+                Formatter<Dictionary<string, int>>.ListExpansionLimit = 1000;
+                Formatter.ListExpansionLimit = 4;
+                var dictionary = new Dictionary<string, int>
+                {
+                    { "zero", 0 },
+                    { "two", 2 },
+                    { "three", 3 },
+                    { "four", 4 },
+                    { "five", 5 },
+                    { "six", 6 },
+                    { "seven", 7 },
+                    { "eight", 8 },
+                    { "nine", 9 },
+                    { "ninety-nine", 99 }
+                };
+
+                var output = dictionary.ToDisplayString();
+
+                output.Should().Contain("zero");
+                output.Should().Contain("0");
+                output.Should().Contain("ninety-nine");
+                output.Should().Contain("99");
             }
-            else
+
+            [Fact]
+            public void Formatting_can_be_chosen_based_on_mime_type()
             {
                 Formatter.Register(
-                    type: typeof(AbstractClass),
-                    formatter: (obj, writer) => writer.Write(obj.GetType().Name.ToUpperInvariant()), mimeType);
-            }
-
-            var instanceOf1 = new ConcreteClass1InheritingAbstractClass();
+                    new PlainTextFormatter<DateTime>((time, c) => c.Writer.Write("plain")));
+                Formatter.Register(
+                    new HtmlFormatter<DateTime>((time, c) => c.Writer.Write("html")));
 
-            instanceOf1.ToDisplayString(mimeType)
-                       .Should()
-                       .Be(instanceOf1.GetType().Name.ToUpperInvariant());
+                var now = DateTime.Now;
 
-            var instanceOf2 = new ConcreteClass2InheritingAbstractClass();
+                now.ToDisplayString(PlainTextFormatter.MimeType).Should().Be("plain");
+                now.ToDisplayString(HtmlFormatter.MimeType).Should().Be("html");
+            }
 
-            instanceOf2.ToDisplayString(mimeType)
-                       .Should()
-                       .Be(instanceOf2.GetType().Name.ToUpperInvariant());
-        }
+            [Theory]
+            [InlineData("text/plain")]
+            [InlineData("text/html")]
+            public void Formatter_sources_can_provide_lazy_registration_of_custom_formatters(string mimeType)
+            {
+                var obj = new TypeWithCustomFormatter();
 
-        public abstract class AbstractClass
-        {
-        };
+                var formatted = obj.ToDisplayString(mimeType);
 
-        public class ConcreteClass1InheritingAbstractClass : AbstractClass
-        {
-        };
+                formatted.Should().Be($"Hello from {nameof(TypeWithCustomFormatter)} using MIME type {mimeType}");
+            }
 
-        public class ConcreteClass2InheritingAbstractClass : AbstractClass
-        {
-        };
+            [TypeFormatterSource(typeof(CustomFormatterSource))]

```

Comment:
I like the idea of allowing types to say "Here's how I want to be formatted!" but I don't love using attributes to do it. Unless we write an analyzer, there's no compiler feedback for specifying the wrong type. 

Any other ideas?